### PR TITLE
Irq rework master

### PR DESCRIPTION
**Pull request by samsor**
Rework of libmetal IRQ API:
 -Separate original and single metal_irq_register() into two independent functions metal_irq_register() and metal_irq_unregister().
 -Change Baremetal and FreeRTOS implementation to replace irq LUT and handler array with dynamically allocated containers.
 -Process optional device parameter for IRQ registration with Baremetal and FreeRTOS implementation.
 -Change Linux irq handling to check poll()'s revents POLLIN|POLLRDNORM instead of POLLIN|POLLERR
 -Improve irq unit tests to use new metal_irq_unregister() and add more coverage, also add Linux coverage.
 -Remove unneedeed irq_enable flag in generic and FreeRTOS irq lib
 -Change Linux implementation to replace irq handler array with dynamically allocated container.

### CODE REVIEW

File: lib/system/linux/device.c [hunk 1]
```diff
@@ -305,7 +308,7 @@ static void metal_uio_dev_irq_ack(struct linux_bus *lbus,
 {
 	(void)lbus;
 	(void)irq;
-	unsigned int irq_info = 1;
+	int irq_info = 1;
 	unsigned int val;
```
Comment (wjliang) [inline lib/system/linux/device.c]: why change from int to "unsigned int", please revert this change as it doesn't look like relevant to this patch series

File: lib/irq.h [hunk 2]
```diff
@@ -60,30 +60,44 @@ typedef int (*metal_irq_handler) (int irq, void *priv);
...
  * @param[in]  drv_id      driver id. It can be used for driver data.
  * @return     0 for success, non-zero on failure
  */
-int metal_irq_register(int irq,
-		       metal_irq_handler irq_handler,
-		       struct metal_device *dev,
-		       void *drv_id);
+int metal_irq_unregister(int irq,
+                         metal_irq_handler irq_handler,
+                         struct metal_device *dev,
```
Comment (wjliang) [inline lib/irq.h]: looks like white space difference.

File: lib/system/linux/device.c (hunk 1 above)
Comment (samsor) [reply lib/system/linux/device.c]: the opposite, changed from unsigned int to int.

File: lib/system/linux/device.c [hunk 3]
```diff
@@ -176,7 +176,7 @@ static int metal_uio_dev_open(struct linux_bus *lbus, struct linux_device *ldev)
 	struct dlist *dlist;
 	int result, i;
 	void *virt;
-	int irq_info;
+	unsigned int irq_info;
```
[the remark this answers is not in the document]
Comment (samsor) [reply lib/system/linux/device.c]: OK

File: lib/irq.h (hunk 2 above)
Comment (samsor) [reply lib/irq.h]: OK
